lightning-block-sync/src/http.rs

```diff
@@ -0,0 +1,765 @@
+use chunked_transfer;
+use serde_json;
+
+use std::convert::TryFrom;
+#[cfg(not(feature = "tokio"))]
+use std::io::Write;
+use std::net::ToSocketAddrs;
+use std::time::Duration;
+
+#[cfg(feature = "tokio")]
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt};
+#[cfg(feature = "tokio")]
+use tokio::net::TcpStream;
+
+#[cfg(not(feature = "tokio"))]
+use std::io::BufRead;
+use std::io::Read;
+#[cfg(not(feature = "tokio"))]
+use std::net::TcpStream;
+
+/// Maximum HTTP message header size in bytes.
+const MAX_HTTP_MESSAGE_HEADER_SIZE: usize = 8192;
+
+/// Maximum HTTP message body size in bytes. Enough for a hex-encoded block in JSON format.
+const MAX_HTTP_MESSAGE_BODY_SIZE: usize = 2 * 4_000_000 + 64;
+
+/// Endpoint for interacting with an HTTP-based API.
+#[derive(Debug)]
+pub struct HttpEndpoint {
+	scheme: Scheme,
+	host: String,
+	port: Option<u16>,
+	path: String,
+}
+
+/// URI scheme compatible with an HTTP endpoint.
+#[derive(Debug)]
+pub enum Scheme {
+	HTTP,
+	HTTPS,
+}
+
+impl HttpEndpoint {
+	/// Creates an endpoint using the HTTP scheme.
+	pub fn insecure_host(host: String) -> Self {
+		Self {
+			scheme: Scheme::HTTP,
+			host,
+			port: None,
+			path: String::from("/"),
+		}
+	}
+
+	/// Creates an endpoint using the HTTPS scheme.
+	pub fn secure_host(host: String) -> Self {
```

Comment:
> Generally, I'd prefer to make unit tests hermetic whenever possible to avoid flakiness with the external world.

I've come to agree with this, like we wouldn't want to require an internet connection to run our tests.

> Having separate tests would be useful, though. I would add a placeholder now to demonstrate that it's currently not supported, but that isn't possible with the current API taking E: ToSocketAddrs.

Hmm should've asked in the call but I wasn't sure what this meant. Separate tests for what? and what does "placeholder" mean?